brain/detail/circuit.h

```diff
@@ -222,12 +189,335 @@ class Circuit::Impl
+
+        Vector3fs output;
+        for (auto gid : gids)
+        {
+            const float x = attributeX[gid - startIdx];
+            const float y = attributeY[gid - startIdx];
+            const float z = attributeZ[gid - startIdx];
+
+            output.push_back(Vector3f(x, y, z));
+        }
+        return output;
+    }
+    virtual size_ts getMTypes(const GIDSet& /*gids*/) const
+    {
+        LBUNIMPLEMENTED;
+        return size_ts();
+    }
+    virtual Strings getMorphologyNames() const
+    {
+        return nodeGroup.getAttribute<std::string>("morphology_file");
+    }
+    virtual size_ts getETypes(const GIDSet& /*gids*/) const
+    {
+        LBUNIMPLEMENTED;
+        return size_ts();
+    }
+    virtual Strings getElectrophysiologyNames() const
+    {
+        LBUNIMPLEMENTED;
+        return Strings();
+    }
+    virtual Quaternionfs getRotations(const GIDSet& gids) const
+    {
+        if (gids.empty())
+            return Quaternionfs();
+
+        const size_t startIdx = *gids.begin();
+        const size_t endIdx = *gids.rbegin() + 1;
+
+        const auto rotationAngleX =
+            nodeGroup.getAttribute<float>("rotation_angle_x", startIdx, endIdx);
+        const auto rotationAngleY =
+            nodeGroup.getAttribute<float>("rotation_angle_y", startIdx, endIdx);
+        const auto rotationAngleZ =
+            nodeGroup.getAttribute<float>("rotation_angle_z", startIdx, endIdx);
+
+        Quaternionfs output;
+
+        for (const auto gid : gids)
+        {
+            const float rX = rotationAngleX[gid - startIdx];
+            const float rY = rotationAngleY[gid - startIdx];
+            const float rZ = rotationAngleZ[gid - startIdx];
+
+            const float cX = std::cos(rX), cY = std::cos(rY), cZ = std::cos(rZ);
+            const float sX = std::sin(rX), sY = std::sin(rY), sZ = std::sin(rZ);
+
+            // These are the values given by multiplying the rotation
+            // matrices for R(X)R(Y)R(Z) i.e. extrinsic rotation around Z
+            // then Y then X
+            vmml::Matrix3f mtx;
+
+            mtx(0, 0) = cY * cZ;
+            mtx(0, 1) = -cY * sZ;
+            mtx(0, 2) = sY;
+            mtx(1, 0) = cZ * sX * sY + cX * sZ;
+            mtx(1, 1) = cX * cZ - sX * sY * sZ;
+            mtx(1, 2) = -cY * sX;
+            mtx(2, 0) = sX * sZ - cX * cZ * sY;
+            mtx(2, 1) = cZ * sX + cX * sY * sZ;
+            mtx(2, 2) = cX * cY;
+
+            output.push_back(Quaternionf(mtx));
+        }
+
+        return output;
+    }
+    virtual Strings getMorphologyNames(const GIDSet& gids) const
+    {
+        if (gids.empty())
+            return Strings();
+
+        const size_t startIdx = *gids.begin();
+        const size_t endIdx = *gids.rbegin() + 1;
+
+        const auto names =
+            nodeGroup.getAttribute<std::string>("morphology_file", startIdx,
+                                                endIdx);
```

Comment:
If this attribute doesn't exist, you have to look up in the CSV file based on the node type. This is exactly the case of ABI's circuits.